File: mobile/ios/LaligaOAuthPlugin.swift
import Foundation
import AuthenticationServices
import Capacitor
import UIKit

/**
 * Puente nativo mínimo para el login social de LALIGA Fantasy.
 *
 * La web entrega solo la URL de autorización. ASWebAuthenticationSession abre
 * el dominio oficial de LALIGA y captura dentro de ESTA sesión el callback
 * `authredirect://...`. Al JavaScript vuelve únicamente esa URL con `code` y
 * `state`; el intercambio por tokens se hace en el backend de LigaLab.
 */
@objc(LaligaOAuthPlugin)
public class LaligaOAuthPlugin: CAPPlugin, CAPBridgedPlugin, ASWebAuthenticationPresentationContextProviding {
    public let identifier = "LaligaOAuthPlugin"
    public let jsName = "LaligaOAuth"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise)
    ]

    private var authSession: ASWebAuthenticationSession?

    @objc func start(_ call: CAPPluginCall) {
        guard let rawUrl = call.getString("url"),
              let url = URL(string: rawUrl),
              let scheme = url.scheme?.lowercased(),
              scheme == "https" else {
            call.reject("La URL de acceso de LALIGA no es válida.")
            return
        }

        guard url.host?.lowercased() == "login.laliga.es" else {
            call.reject("Solo se permite abrir el dominio oficial de acceso de LALIGA.")
            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let self else {
                call.reject("No se pudo iniciar el acceso.")
                return
            }

            let session = ASWebAuthenticationSession(
                url: url,
                callback: .customScheme("authredirect")
            ) { [weak self] callbackURL, error in
                defer { self?.authSession = nil }

                if let authError = error as? ASWebAuthenticationSessionError,
                   authError.code == .canceledLogin {
                    call.reject("Acceso cancelado.")
                    return
                }

                if let error {
                    call.reject("No se pudo completar el acceso: \(error.localizedDescription)")
                    return
                }

                guard let callbackURL,
                      callbackURL.scheme?.lowercased() == "authredirect",
                      callbackURL.host?.lowercased() == "com.lfp.laligafantasy" else {
                    call.reject("LALIGA no devolvió el callback esperado.")
                    return
                }

                call.resolve(["callbackUrl": callbackURL.absoluteString])
            }

            session.presentationContextProvider = self
            // Conserva la sesión del proveedor para que Google/Apple/Facebook
            // puedan reutilizar el login que ya tenga el usuario en iOS.
            session.prefersEphemeralWebBrowserSession = false
            self.authSession = session

            guard session.start() else {
                self.authSession = nil
                call.reject("iOS no pudo abrir la sesión de acceso.")
                return
            }
        }
    }

    public func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        if let window = bridge?.viewController?.view.window {
            return window
        }

        let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        if let window = scenes.flatMap({ $0.windows }).first(where: { $0.isKeyWindow }) {
            return window
        }

        return ASPresentationAnchor()
    }
}

File: mobile/android/LaligaOAuthPlugin.kt
package com.inigo.ligalab

import android.content.Intent
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * Puente nativo minimo para el login social de LALIGA Fantasy en Android.
 *
 * Es el gemelo de `mobile/ios/LaligaOAuthPlugin.swift` y expone EXACTAMENTE el
 * mismo contrato (`LaligaOAuth.start({url}) -> {callbackUrl}`), porque el lado
 * web no distingue plataforma: `app/fantasy/mobile-auth.ts` solo busca
 * `Capacitor.Plugins.LaligaOAuth`. Dos contratos distintos obligarian a meter
 * ramas por sistema operativo en la web, que es justo lo que no hace falta.
 *
 * ── Por que existe esto y no basta la web ───────────────────────────────────
 * LALIGA solo acepta como redirect el esquema nativo
 * `authredirect://com.lfp.laligafantasy`. Un navegador no puede recibir ese
 * callback, asi que quien creo su cuenta de LALIGA con Google, Apple o Facebook
 * no tiene forma de entrar desde la web sin pegar un token a mano. Dentro de una
 * app que declara ese esquema, si.
 *
 * ── Que NO pasa por aqui ────────────────────────────────────────────────────
 * Ni la contraseña ni el `code_verifier` de PKCE. Al JavaScript vuelve solo la
 * URL de callback con `code` y `state`; el intercambio por tokens lo hace el
 * backend, que guarda el verifier en una cookie HttpOnly que este plugin nunca
 * ve. Aqui solo se abre una pantalla y se recoge una URL.
 */
@CapacitorPlugin(name = "LaligaOAuth")
class LaligaOAuthPlugin : Plugin() {

    /** La llamada en curso. Solo puede haber una: dos logins a la vez no existen. */
    private var pendiente: PluginCall? = null

    /**
     * Se ha abierto la pestaña y todavia no ha vuelto el callback.
     *
     * Es lo que permite distinguir «el usuario ha cerrado la pestaña» de «ha
     * terminado bien»: Android no avisa de que se cierre un Custom Tab, asi que
     * la unica señal es volver a `onResume` sin haber recibido el `Intent`.
     */
    private var esperandoRedireccion = false

    /**
     * La actividad ha llegado a pausarse desde que se abrio la pestaña.
     *
     * Sin esta segunda bandera el plugin se cancela solo en algunos
     * dispositivos: lanzar un Custom Tab puede disparar un `onResume` de la
     * actividad ANTES de que llegue el `onPause`, y ese resume se leeria como
     * «el usuario ha cerrado la pestaña» cuando en realidad todavia no la ha
     * visto. Solo cuenta como cancelacion volver despues de haber salido.
     */
    private var haSalido = false

    @PluginMethod
    fun start(call: PluginCall) {
        val cruda = call.getString("url")
        if (cruda.isNullOrBlank()) {
            call.reject("La URL de acceso de LALIGA no es válida.")
            return
        }

        val url = try {
            Uri.parse(cruda)
        } catch (_: Exception) {
            call.reject("La URL de acceso de LALIGA no es válida.")
            return
        }

        /*
         * La lista blanca es la misma que en iOS, y no es una formalidad: esta
         * URL la construye el servidor, pero si algun dia llegara manipulada
         * abririamos una pantalla de acceso ajena con la apariencia de la
         * nuestra. Lo que no es HTTPS contra el dominio oficial, no se abre.
         */
        if (!"https".equals(url.scheme, ignoreCase = true)) {
            call.reject("La URL de acceso de LALIGA no es válida.")
            return
        }
        if (!"login.laliga.es".equals(url.host, ignoreCase = true)) {
            call.reject("Solo se permite abrir el dominio oficial de acceso de LALIGA.")
            return
        }

        // Una llamada nueva cancela la anterior en vez de dejarla colgada: si no,
        // su promesa no se resolveria nunca y la pantalla se quedaria esperando.
        pendiente?.reject("Acceso cancelado.")

        call.setKeepAlive(true)
        pendiente = call
        esperandoRedireccion = true
        haSalido = false

        try {
            val pestana = CustomTabsIntent.Builder()
                .setShowTitle(true)
                .build()
            pestana.intent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
            pestana.launchUrl(activity, url)
        } catch (error: Exception) {
            // Sin navegador compatible con Custom Tabs no hay nada que abrir.
            limpiar()
            call.reject("Android no pudo abrir la pantalla de acceso: ${error.message ?: "sin navegador disponible"}")
        }
    }

    /**
     * Llega el `authredirect://com.lfp.laligafantasy?...`.
     *
     * Requiere que la actividad declare ese esquema y use `singleTask`; con el
     * modo por defecto Android abriria una instancia nueva y este metodo no se
     * llamaria nunca. Ver `mobile/android/README.md`.
     */
    override fun handleOnNewIntent(intent: Intent) {
        super.handleOnNewIntent(intent)

        val datos = intent.data ?: return
        if (!"authredirect".equals(datos.scheme, ignoreCase = true)) return
        if (!"com.lfp.laligafantasy".equals(datos.host, ignoreCase = true)) return

        val call = pendiente ?: return
        // Antes de resolver: si no se baja la bandera aqui, el `onResume` que
        // viene justo despues rechazaria la llamada por «cancelada».
        limpiar()

        val respuesta = JSObject()
        respuesta.put("callbackUrl", datos.toString())
        call.resolve(respuesta)
    }

    /**
     * Volver aqui con una llamada viva y sin callback significa que el usuario
     * cerro la pestaña. Sin esto, la promesa se quedaria pendiente para siempre
     * y el boton de la pantalla no volveria de «Entrando…».
     */
    override fun handleOnPause() {
        super.handleOnPause()
        if (esperandoRedireccion) haSalido = true
    }

    override fun handleOnResume() {
        super.handleOnResume()
        if (!esperandoRedireccion || !haSalido) return

        val call = pendiente ?: return
        limpiar()
        call.reject("Acceso cancelado.")
    }

    private fun limpiar() {
        pendiente = null
        esperandoRedireccion = false
        haSalido = false
    }
}
